jl4/tests/SemanticTokens.hs

```diff
@@ -0,0 +1,71 @@
+module SemanticTokens where
+
+import Base
+
+import qualified Base.Text as Text
+import qualified LSP.Core.Shake as Shake
+import LSP.L4.Oneshot (oneshotL4ActionAndErrors)
+import qualified LSP.L4.Rules as Rules
+import LSP.SemanticTokens
+import Language.LSP.Protocol.Types
+import Optics
+import System.FilePath
+import Test.Hspec
+import Test.Hspec.Golden
+
+semanticTokenTests :: [FilePath] -> FilePath -> Spec
+semanticTokenTests files root = do
+  forM_ files $ \inputFile -> do
+    let
+      testCase = makeRelative root inputFile
+    let
+      goldenDir = takeDirectory inputFile </> "tests"
+    it testCase $ do
+      semanticTokenGolden goldenDir inputFile
+
+semanticTokenGolden :: String -> FilePath -> IO (Golden Text)
+semanticTokenGolden dir inputFile = do
+  (_errs, moutput) <- oneshotL4ActionAndErrors inputFile \nfp -> do
+    let
+      uri = normalizedFilePathToUri nfp
+    _ <- Shake.addVirtualFileFromFS nfp
+    _ <- Shake.use Rules.TypeCheck uri
+    lexerTokens <- Shake.use Rules.LexerSemanticTokens uri
+    parserTokens <- Shake.use Rules.ParserSemanticTokens uri
+    checkedTokens <- Shake.use Rules.TypeCheckedSemanticTokens uri
+    pure (lexerTokens, parserTokens, checkedTokens)
+  let
+    output_ :: Text -> Maybe [SemanticToken] -> Text
+    output_ phase mToks =
+      "Phase: " <> phase <> "\n" <>
+      case mToks of
+        Nothing -> "No Tokens for this phase"
```

Comment:
that's not what this means, it means that the entire thing failed. I think we should rather assert that this is `Just` up top and then just not have this case here.

Reply:
No, for example when a file doesn't typecheck, then this phase may reasonably not produce any tokens.
None of the tests currently test the case of partially correct files, but I would like to add some.

Reply:
But doesn’t this already return Nothing when any of the phases fail?

Reply:
No, the output is `moutput :: (Maybe [SemanticToken], Maybe [SemanticToken],
 Maybe [SemanticToken])`, I want to explicitly know when a specific phase fails and record it in the test file.